In [566]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [567]:
#let us first load the train dataset with the name df
df=pd.read_csv('Train_cars.csv')
df.head()

Name    Location  Year  Kilometers_Driven Fuel_Type  \
0     Maruti 800 Std BSIII   Bangalore  2012              33715    Diesel   
1      Honda City 1.5 V AT       Delhi  2013              73977    Diesel   
2        Honda CR-V RVi MT       Delhi  2007              16460       CNG   
3  Skoda Yeti Ambition 4X2     Kolkata  2009              35048    Petrol   
4    Hyundai i10 Magna LPG  Coimbatore  2013              41656    Petrol   

  Transmission Owner_Type     Mileage   Engine     Power  Seats   New_Price  \
0       Manual      First  16.46 kmpl  1248 CC   100 bhp    5.0  43.65 Lakh   
1       Manual      First   18.6 kmpl  2993 CC    61 bhp    4.0         NaN   
2       Manual      Third   15.3 kmpl  1599 CC   140 bhp    5.0         NaN   
3       Manual      First  25.47 kmpl  1196 CC  86.8 bhp    5.0         NaN   
4    Automatic      First  19.12 kmpl  1999 CC   115 bhp    7.0  13.26 Lakh   

   Price  
0  25.53  
1   4.58  
2   2.37  
3   1.20  
4  33.04

In [568]:
df_test = pd.read_csv('Test_cars.csv')
df_test.head()
#let us load the testing dataset with name df_test

Name   Location  Year  Kilometers_Driven Fuel_Type  \
0     Ford Figo Diesel Titanium      Delhi  2010              48928    Diesel   
1                Maruti SX4 VDI     Jaipur  2011              47428    Diesel   
2   Hyundai Creta 1.6 SX Option  Bangalore  2014              28984    Petrol   
3  Jeep Compass 2.0 Limited 4X4      Kochi  2017              46735    Petrol   
4        Maruti Swift Dzire VDI      Kochi  2015              19975    Petrol   

  Transmission Owner_Type     Mileage   Engine      Power  Seats   New_Price  
0    Automatic      First   17.8 kmpl  1248 CC    258 bhp    7.0         NaN  
1    Automatic      First   23.9 kmpl  2498 CC    183 bhp    5.0   4.33 Lakh  
2    Automatic      First  19.59 kmpl  2494 CC    203 bhp    5.0  71.15 Lakh  
3       Manual      First   23.9 kmpl  1498 CC    215 bhp    5.0         NaN  
4    Automatic      First  26.6 km/kg  2993 CC  147.8 bhp    7.0         NaN

In [569]:
#printing the information about the both datasets
print(df.info())
print(df_test.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 22500 entries, 0 to 22499
Data columns (total 13 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Name               22500 non-null  object 
 1   Location           22500 non-null  object 
 2   Year               22500 non-null  int64  
 3   Kilometers_Driven  22500 non-null  int64  
 4   Fuel_Type          22500 non-null  object 
 5   Transmission       22500 non-null  object 
 6   Owner_Type         22500 non-null  object 
 7   Mileage            22408 non-null  object 
 8   Engine             22463 non-null  object 
 9   Power              22415 non-null  object 
 10  Seats              22408 non-null  float64
 11  New_Price          3881 non-null   object 
 12  Price              22500 non-null  float64
dtypes: float64(2), int64(2), object(9)
memory usage: 2.2+ MB
None
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 52500 entries, 0 to 52499
Data columns (total 12 colum

In [570]:
#printing the data types of both the data sets
print(df.dtypes)
print(df_test.dtypes)

Name                  object
Location              object
Year                   int64
Kilometers_Driven      int64
Fuel_Type             object
Transmission          object
Owner_Type            object
Mileage               object
Engine                object
Power                 object
Seats                float64
New_Price             object
Price                float64
dtype: object
Name                  object
Location              object
Year                   int64
Kilometers_Driven      int64
Fuel_Type             object
Transmission          object
Owner_Type            object
Mileage               object
Engine                object
Power                 object
Seats                float64
New_Price             object
dtype: object


In [571]:
print('Rows:' , df.shape[0])
print('Columns:' , df.shape[1])
print('Rows of test:' , df_test.shape[0])
print('Columns of test:' , df_test.shape[1])

Rows: 22500
Columns: 13
Rows of test: 52500
Columns of test: 12


In [572]:
df.describe()

Year  Kilometers_Driven         Seats         Price
count  22500.000000       22500.000000  22408.000000  22500.000000
mean    2011.819378       49109.527556      5.352330      6.359158
std        2.663807       21746.602938      0.842943      7.302764
min     2001.000000         171.000000      4.000000      0.440000
25%     2010.000000       35833.250000      5.000000      2.290000
50%     2012.000000       47991.000000      5.000000      3.510000
75%     2014.000000       59696.500000      5.000000      6.590000
max     2018.000000      253494.000000      8.000000     71.600000

In [573]:
df.describe(include=object)

Name Location Fuel_Type Transmission Owner_Type  \
count                    22500    22500     22500        22500      22500   
unique                    1204       11         5            2          4   
top     Maruti Wagon R LXI CNG   Mumbai    Petrol       Manual      First   
freq                       142     4387     10155        18046      20257   

           Mileage  Engine      Power  New_Price  
count        22408   22463      22415       3881  
unique         442     146        372        528  
top     26.6 km/kg  998 CC  58.16 bhp  8.61 Lakh  
freq          1310    2293       1334        980

In [574]:
df_test.describe()

Year  Kilometers_Driven         Seats
count  52500.000000       52500.000000  52254.000000
mean    2011.821771       48751.882171      5.357867
std        2.695324       21512.403003      0.849945
min     2000.000000         171.000000      4.000000
25%     2010.000000       35640.500000      5.000000
50%     2012.000000       47817.000000      5.000000
75%     2014.000000       59492.000000      5.000000
max     2018.000000      255022.000000      8.000000

In [575]:
df_test.describe(include=object)

Name Location Fuel_Type Transmission Owner_Type  \
count                    52500    52500     52500        52500      52500   
unique                    1431       11         5            2          4   
top     Maruti Wagon R LXI CNG   Mumbai    Diesel       Manual      First   
freq                       307    10045     23563        42061      47195   

           Mileage  Engine      Power  New_Price  
count        52251   52431      52268       9185  
unique         442     146        372        537  
top     26.6 km/kg  998 CC  58.16 bhp  8.61 Lakh  
freq          2902    5105       3216       2374

In [576]:
print(df['Location'].unique())
print(df['Year'].unique())
print(df['Fuel_Type'].unique())
print(df['Transmission'].unique())
print(df['Owner_Type'].unique())

['Bangalore' 'Delhi' 'Kolkata' 'Coimbatore' 'Pune' 'Hyderabad' 'Mumbai'
 'Kochi' 'Chennai' 'Jaipur' 'Ahmedabad']
[2012 2013 2007 2009 2011 2010 2016 2017 2014 2015 2008 2004 2003 2006
 2005 2018 2002 2001]
['Diesel' 'CNG' 'Petrol' 'LPG' 'Electric']
['Manual' 'Automatic']
['First' 'Third' 'Second' 'Fourth & Above']


In [577]:
print(df_test['Location'].unique())
print(df_test['Year'].unique())
print(df_test['Fuel_Type'].unique())
print(df_test['Transmission'].unique())
print(df_test['Owner_Type'].unique())

['Delhi' 'Jaipur' 'Bangalore' 'Kochi' 'Kolkata' 'Pune' 'Hyderabad'
 'Mumbai' 'Ahmedabad' 'Chennai' 'Coimbatore']
[2010 2011 2014 2017 2015 2009 2005 2012 2007 2013 2004 2008 2001 2016
 2003 2006 2018 2002 2000]
['Diesel' 'Petrol' 'CNG' 'LPG' 'Electric']
['Automatic' 'Manual']
['First' 'Second' 'Third' 'Fourth & Above']


Now as we can see that Mileage , Engine and Power are object column but it should be a numeric column so now we will chnage the datatype of these column to appropriate datatypes for both training and testing datatypes

In [578]:
df_test['Mileage']

0         17.8 kmpl
1         23.9 kmpl
2        19.59 kmpl
3         23.9 kmpl
4        26.6 km/kg
            ...    
52495    16.09 kmpl
52496     12.6 kmpl
52497     22.0 kmpl
52498    10.98 kmpl
52499     14.0 kmpl
Name: Mileage, Length: 52500, dtype: object

In [579]:
df_test[['Mileage','kmpl']]=df_test['Mileage'].str.split(expand=True)

In [580]:
df_test['Mileage']=df_test['Mileage'].astype(float)

In [581]:
df['Mileage']

0         16.46 kmpl
1          18.6 kmpl
2          15.3 kmpl
3         25.47 kmpl
4         19.12 kmpl
            ...     
22495      21.1 kmpl
22496    33.44 km/kg
22497     26.21 kmpl
22498      17.8 kmpl
22499     26.59 kmpl
Name: Mileage, Length: 22500, dtype: object

In [582]:
df[['Mileage','kmpl']]=df['Mileage'].str.split(expand=True)
df['Mileage']=df['Mileage'].astype(float)

In [583]:
df['Mileage']

0        16.46
1        18.60
2        15.30
3        25.47
4        19.12
         ...  
22495    21.10
22496    33.44
22497    26.21
22498    17.80
22499    26.59
Name: Mileage, Length: 22500, dtype: float64

Mileage column in both training and testing have been changed to float

In [584]:
df_test['Engine'] 

0        1248 CC
1        2498 CC
2        2494 CC
3        1498 CC
4        2993 CC
          ...   
52495    1248 CC
52496    1591 CC
52497    2993 CC
52498    1995 CC
52499    1798 CC
Name: Engine, Length: 52500, dtype: object

In [585]:
df_test[['Engine','CC']] = df_test['Engine'].str.split(expand =True)
df_test['Engine']=df_test['Engine'].astype(float)
df_test['Engine']

0        1248.0
1        2498.0
2        2494.0
3        1498.0
4        2993.0
          ...  
52495    1248.0
52496    1591.0
52497    2993.0
52498    1995.0
52499    1798.0
Name: Engine, Length: 52500, dtype: float64

In [586]:
df['Engine']

0        1248 CC
1        2993 CC
2        1599 CC
3        1196 CC
4        1999 CC
          ...   
22495    1368 CC
22496    1197 CC
22497    1298 CC
22498    1198 CC
22499    1248 CC
Name: Engine, Length: 22500, dtype: object

In [587]:
df[['Engine','CC']] = df['Engine'].str.split(expand =True)
df['Engine']=df['Engine'].astype(float)

In [588]:
df['Engine']

0        1248.0
1        2993.0
2        1599.0
3        1196.0
4        1999.0
          ...  
22495    1368.0
22496    1197.0
22497    1298.0
22498    1198.0
22499    1248.0
Name: Engine, Length: 22500, dtype: float64

Enginee column in both training and testing have been changed to float

In train and test we have split the Power column now we will treat the values which have 'null' as one of the entries

In [589]:
df_test['Power']

0          258 bhp
1          183 bhp
2          203 bhp
3          215 bhp
4        147.8 bhp
           ...    
52495     null bhp
52496     85.8 bhp
52497    174.3 bhp
52498      112 bhp
52499    488.1 bhp
Name: Power, Length: 52500, dtype: object

In [590]:
df_test[['Power','bhp']] = df_test['Power'].str.split(expand =True)

In [591]:
df_test['Power'].replace('null',np.NaN,inplace=True)

In [592]:
df_test['Power']=df_test['Power'].astype(float)

In [593]:
df['Power']

0          100 bhp
1           61 bhp
2          140 bhp
3         86.8 bhp
4          115 bhp
           ...    
22495    100.6 bhp
22496          NaN
22497    58.16 bhp
22498       75 bhp
22499    201.1 bhp
Name: Power, Length: 22500, dtype: object

In [594]:
df[['Power','bhp']] = df['Power'].str.split(expand =True)

In [595]:
df['Power'].replace('null',np.NaN,inplace=True)

In [596]:
df['Power']=df['Power'].astype(float)

Power column in both train and test have been converted to float

In [597]:
print(df.dtypes) 
print(df_test.dtypes)

Name                  object
Location              object
Year                   int64
Kilometers_Driven      int64
Fuel_Type             object
Transmission          object
Owner_Type            object
Mileage              float64
Engine               float64
Power                float64
Seats                float64
New_Price             object
Price                float64
kmpl                  object
CC                    object
bhp                   object
dtype: object
Name                  object
Location              object
Year                   int64
Kilometers_Driven      int64
Fuel_Type             object
Transmission          object
Owner_Type            object
Mileage              float64
Engine               float64
Power                float64
Seats                float64
New_Price             object
kmpl                  object
CC                    object
bhp                   object
dtype: object


Checking for null value percentage

In [598]:
(df.isnull().sum()/len(df))*100

Name                  0.000000
Location              0.000000
Year                  0.000000
Kilometers_Driven     0.000000
Fuel_Type             0.000000
Transmission          0.000000
Owner_Type            0.000000
Mileage               0.408889
Engine                0.164444
Power                 3.542222
Seats                 0.408889
New_Price            82.751111
Price                 0.000000
kmpl                  0.408889
CC                    0.164444
bhp                   0.377778
dtype: float64

The New_Price column has 82.75% null value so we will drop this column from df dataframe

In [599]:
df.drop('New_Price',axis=1,inplace=True)

In [600]:
(df_test.isnull().sum()/len(df_test))*100

Name                  0.000000
Location              0.000000
Year                  0.000000
Kilometers_Driven     0.000000
Fuel_Type             0.000000
Transmission          0.000000
Owner_Type            0.000000
Mileage               0.474286
Engine                0.131429
Power                 3.497143
Seats                 0.468571
New_Price            82.504762
kmpl                  0.474286
CC                    0.131429
bhp                   0.441905
dtype: float64

The New_Price column has 82.50% null value so we will drop this column from df dataframe

In [601]:
df_test.drop('New_Price',axis=1,inplace=True)

In [602]:
df.head(2)

Name   Location  Year  Kilometers_Driven Fuel_Type  \
0  Maruti 800 Std BSIII  Bangalore  2012              33715    Diesel   
1   Honda City 1.5 V AT      Delhi  2013              73977    Diesel   

  Transmission Owner_Type  Mileage  Engine  Power  Seats  Price  kmpl  CC  bhp  
0       Manual      First    16.46  1248.0  100.0    5.0  25.53  kmpl  CC  bhp  
1       Manual      First    18.60  2993.0   61.0    4.0   4.58  kmpl  CC  bhp

In [603]:
df_test.head(2)

Name Location  Year  Kilometers_Driven Fuel_Type  \
0  Ford Figo Diesel Titanium    Delhi  2010              48928    Diesel   
1             Maruti SX4 VDI   Jaipur  2011              47428    Diesel   

  Transmission Owner_Type  Mileage  Engine  Power  Seats  kmpl  CC  bhp  
0    Automatic      First     17.8  1248.0  258.0    7.0  kmpl  CC  bhp  
1    Automatic      First     23.9  2498.0  183.0    5.0  kmpl  CC  bhp

Let us drop some more insignificant columns from both training and testing dataset, here we will drop column 'kmpl','CC','bhpm'

In [604]:
df.drop(columns=['kmpl','CC','bhp'] , axis=1 , inplace=True)
df_test.drop(columns=['kmpl','CC','bhp'] , axis=1 , inplace=True)

In [605]:
df.head(2)

Name   Location  Year  Kilometers_Driven Fuel_Type  \
0  Maruti 800 Std BSIII  Bangalore  2012              33715    Diesel   
1   Honda City 1.5 V AT      Delhi  2013              73977    Diesel   

  Transmission Owner_Type  Mileage  Engine  Power  Seats  Price  
0       Manual      First    16.46  1248.0  100.0    5.0  25.53  
1       Manual      First    18.60  2993.0   61.0    4.0   4.58

In [606]:
df_test.head(2)

Name Location  Year  Kilometers_Driven Fuel_Type  \
0  Ford Figo Diesel Titanium    Delhi  2010              48928    Diesel   
1             Maruti SX4 VDI   Jaipur  2011              47428    Diesel   

  Transmission Owner_Type  Mileage  Engine  Power  Seats  
0    Automatic      First     17.8  1248.0  258.0    7.0  
1    Automatic      First     23.9  2498.0  183.0    5.0

In [607]:
df['Transmission'].unique()

array(['Manual', 'Automatic'], dtype=object)

In [608]:
df= pd.get_dummies(data=df,columns=['Transmission'],drop_first=True)
df.head(2)

Name   Location  Year  Kilometers_Driven Fuel_Type  \
0  Maruti 800 Std BSIII  Bangalore  2012              33715    Diesel   
1   Honda City 1.5 V AT      Delhi  2013              73977    Diesel   

  Owner_Type  Mileage  Engine  Power  Seats  Price  Transmission_Manual  
0      First    16.46  1248.0  100.0    5.0  25.53                    1  
1      First    18.60  2993.0   61.0    4.0   4.58                    1

In [609]:
df['Owner_Type'].unique()

array(['First', 'Third', 'Second', 'Fourth & Above'], dtype=object)

In [610]:
df['Owner_Type']= df['Owner_Type'].map( {'First':1,'Third':3,'Second':2,
           'Fourth & Above':4})
df.head(2)

Name   Location  Year  Kilometers_Driven Fuel_Type  \
0  Maruti 800 Std BSIII  Bangalore  2012              33715    Diesel   
1   Honda City 1.5 V AT      Delhi  2013              73977    Diesel   

   Owner_Type  Mileage  Engine  Power  Seats  Price  Transmission_Manual  
0           1    16.46  1248.0  100.0    5.0  25.53                    1  
1           1    18.60  2993.0   61.0    4.0   4.58                    1

In [611]:
df['Fuel_Type'].unique()

array(['Diesel', 'CNG', 'Petrol', 'LPG', 'Electric'], dtype=object)

In [612]:
df['Fuel_Type']= df['Fuel_Type'].map( {'Diesel':1,'CNG':2,'Petrol':3,
           'LPG':4 , 'Electric':5})
df.head(2)

Name   Location  Year  Kilometers_Driven  Fuel_Type  \
0  Maruti 800 Std BSIII  Bangalore  2012              33715          1   
1   Honda City 1.5 V AT      Delhi  2013              73977          1   

   Owner_Type  Mileage  Engine  Power  Seats  Price  Transmission_Manual  
0           1    16.46  1248.0  100.0    5.0  25.53                    1  
1           1    18.60  2993.0   61.0    4.0   4.58                    1

In [613]:
df.dtypes

Name                    object
Location                object
Year                     int64
Kilometers_Driven        int64
Fuel_Type                int64
Owner_Type               int64
Mileage                float64
Engine                 float64
Power                  float64
Seats                  float64
Price                  float64
Transmission_Manual      uint8
dtype: object

In [614]:
df_test['Transmission'].unique()

array(['Automatic', 'Manual'], dtype=object)

In [615]:
df_test= pd.get_dummies(data=df_test,columns=['Transmission'],drop_first=True)
df_test.head(2)

Name Location  Year  Kilometers_Driven Fuel_Type  \
0  Ford Figo Diesel Titanium    Delhi  2010              48928    Diesel   
1             Maruti SX4 VDI   Jaipur  2011              47428    Diesel   

  Owner_Type  Mileage  Engine  Power  Seats  Transmission_Manual  
0      First     17.8  1248.0  258.0    7.0                    0  
1      First     23.9  2498.0  183.0    5.0                    0

In [616]:
df_test['Owner_Type'].unique()

array(['First', 'Second', 'Third', 'Fourth & Above'], dtype=object)

In [617]:
df_test['Owner_Type']= df_test['Owner_Type'].map( {'First':1,'Third':3,'Second':2,
           'Fourth & Above':4})
df_test.head(2)

Name Location  Year  Kilometers_Driven Fuel_Type  \
0  Ford Figo Diesel Titanium    Delhi  2010              48928    Diesel   
1             Maruti SX4 VDI   Jaipur  2011              47428    Diesel   

   Owner_Type  Mileage  Engine  Power  Seats  Transmission_Manual  
0           1     17.8  1248.0  258.0    7.0                    0  
1           1     23.9  2498.0  183.0    5.0                    0

In [618]:
df_test['Fuel_Type'].unique()

array(['Diesel', 'Petrol', 'CNG', 'LPG', 'Electric'], dtype=object)

In [619]:
df_test['Fuel_Type']= df_test['Fuel_Type'].map( {'Diesel':1,'CNG':2,'Petrol':3,
           'LPG':4 , 'Electric':5})
df_test.head(2)

Name Location  Year  Kilometers_Driven  Fuel_Type  \
0  Ford Figo Diesel Titanium    Delhi  2010              48928          1   
1             Maruti SX4 VDI   Jaipur  2011              47428          1   

   Owner_Type  Mileage  Engine  Power  Seats  Transmission_Manual  
0           1     17.8  1248.0  258.0    7.0                    0  
1           1     23.9  2498.0  183.0    5.0                    0

In [620]:
df_test.dtypes

Name                    object
Location                object
Year                     int64
Kilometers_Driven        int64
Fuel_Type                int64
Owner_Type               int64
Mileage                float64
Engine                 float64
Power                  float64
Seats                  float64
Transmission_Manual      uint8
dtype: object

In [621]:
location_dict = df['Location'].value_counts(normalize=True).to_dict()
location_dict

{'Mumbai': 0.19497777777777778,
 'Hyderabad': 0.1701777777777778,
 'Kolkata': 0.15902222222222223,
 'Chennai': 0.1039111111111111,
 'Jaipur': 0.07533333333333334,
 'Delhi': 0.06555555555555556,
 'Bangalore': 0.057866666666666663,
 'Kochi': 0.0504,
 'Coimbatore': 0.04724444444444444,
 'Pune': 0.044977777777777776,
 'Ahmedabad': 0.030533333333333332}

In [622]:
df['Location'] = df['Location'].map(location_dict)
df.head()

Name  Location  Year  Kilometers_Driven  Fuel_Type  \
0     Maruti 800 Std BSIII  0.057867  2012              33715          1   
1      Honda City 1.5 V AT  0.065556  2013              73977          1   
2        Honda CR-V RVi MT  0.065556  2007              16460          2   
3  Skoda Yeti Ambition 4X2  0.159022  2009              35048          3   
4    Hyundai i10 Magna LPG  0.047244  2013              41656          3   

   Owner_Type  Mileage  Engine  Power  Seats  Price  Transmission_Manual  
0           1    16.46  1248.0  100.0    5.0  25.53                    1  
1           1    18.60  2993.0   61.0    4.0   4.58                    1  
2           3    15.30  1599.0  140.0    5.0   2.37                    1  
3           1    25.47  1196.0   86.8    5.0   1.20                    1  
4           1    19.12  1999.0  115.0    7.0  33.04                    0

In [623]:
df['Name'].value_counts()

Maruti Wagon R LXI CNG                     142
Maruti Swift VDI                           133
Honda City 1.5 S MT                        131
Hyundai i10 Sportz                         114
BMW 3 Series 320d                          111
                                          ... 
BMW 5 Series 2013-2017 525d Luxury Line      1
Honda Amaze S i-DTEC                         1
Volkswagen Beetle 2.0                        1
Toyota Camry A/T                             1
Toyota Etios VXD                             1
Name: Name, Length: 1204, dtype: int64

In [624]:
Name_dict = round(df.groupby(['Name'])['Price'].mean(),0).to_dict()
Name_dict

{'Audi A3 35 TDI Attraction': 5.0,
 'Audi A3 35 TDI Premium': 10.0,
 'Audi A4 1.8 TFSI': 6.0,
 'Audi A4 2.0 TDI': 7.0,
 'Audi A4 2.0 TDI 177 Bhp Premium Plus': 8.0,
 'Audi A4 2.0 TDI 177 Bhp Technology Edition': 7.0,
 'Audi A4 2.0 TDI Multitronic': 7.0,
 'Audi A4 2.0 TDI Premium Sport Limited Edition': 6.0,
 'Audi A4 3.0 TDI Quattro': 12.0,
 'Audi A4 35 TDI Premium': 8.0,
 'Audi A4 35 TDI Premium Plus': 5.0,
 'Audi A4 35 TDI Technology': 8.0,
 'Audi A4 35 TDI Technology Edition': 3.0,
 'Audi A4 New 2.0 TDI Multitronic': 6.0,
 'Audi A6 2.0 TDI Design Edition': 8.0,
 'Audi A6 2.7 TDI': 10.0,
 'Audi A6 2.8 FSI': 3.0,
 'Audi A6 2011-2015 2.0 TDI': 3.0,
 'Audi A6 2011-2015 2.0 TDI Premium Plus': 7.0,
 'Audi A6 2011-2015 2.0 TDI Technology': 11.0,
 'Audi A6 2011-2015 35 TDI Technology': 5.0,
 'Audi A6 2011-2015 35 TFSI Technology': 6.0,
 'Audi A6 3.0 TDI quattro': 8.0,
 'Audi A6 35 TDI Matrix': 5.0,
 'Audi A6 35 TFSI Matrix': 3.0,
 'Audi A8 L 3.0 TDI quattro': 4.0,
 'Audi Q3 2.0 TDI': 6.0,
 

In [625]:
sum(Name_dict.values())

8236.0

In [626]:
df['Name'] = (df['Name'].map(Name_dict))/sum(Name_dict.values())

In [627]:
df.head()

Name  Location  Year  Kilometers_Driven  Fuel_Type  Owner_Type  \
0  0.000729  0.057867  2012              33715          1           1   
1  0.000850  0.065556  2013              73977          1           1   
2  0.000850  0.065556  2007              16460          2           3   
3  0.001457  0.159022  2009              35048          3           1   
4  0.000971  0.047244  2013              41656          3           1   

   Mileage  Engine  Power  Seats  Price  Transmission_Manual  
0    16.46  1248.0  100.0    5.0  25.53                    1  
1    18.60  2993.0   61.0    4.0   4.58                    1  
2    15.30  1599.0  140.0    5.0   2.37                    1  
3    25.47  1196.0   86.8    5.0   1.20                    1  
4    19.12  1999.0  115.0    7.0  33.04                    0

In [628]:
df.dtypes

Name                   float64
Location               float64
Year                     int64
Kilometers_Driven        int64
Fuel_Type                int64
Owner_Type               int64
Mileage                float64
Engine                 float64
Power                  float64
Seats                  float64
Price                  float64
Transmission_Manual      uint8
dtype: object

In [629]:
location_dict_test = df_test['Location'].value_counts(normalize=True).to_dict()
location_dict_test

{'Mumbai': 0.19133333333333333,
 'Hyderabad': 0.1721904761904762,
 'Kolkata': 0.15148571428571428,
 'Chennai': 0.10459047619047619,
 'Jaipur': 0.08036190476190476,
 'Delhi': 0.06588571428571428,
 'Bangalore': 0.05851428571428571,
 'Kochi': 0.04933333333333333,
 'Coimbatore': 0.04836190476190476,
 'Pune': 0.04609523809523809,
 'Ahmedabad': 0.03184761904761905}

In [630]:
df_test['Location'] = df_test['Location'].map(location_dict_test)
df_test.head()

Name  Location  Year  Kilometers_Driven  Fuel_Type  \
0     Ford Figo Diesel Titanium  0.065886  2010              48928          1   
1                Maruti SX4 VDI  0.080362  2011              47428          1   
2   Hyundai Creta 1.6 SX Option  0.058514  2014              28984          3   
3  Jeep Compass 2.0 Limited 4X4  0.049333  2017              46735          3   
4        Maruti Swift Dzire VDI  0.049333  2015              19975          3   

   Owner_Type  Mileage  Engine  Power  Seats  Transmission_Manual  
0           1    17.80  1248.0  258.0    7.0                    0  
1           1    23.90  2498.0  183.0    5.0                    0  
2           1    19.59  2494.0  203.0    5.0                    0  
3           1    23.90  1498.0  215.0    5.0                    1  
4           1    26.60  2993.0  147.8    7.0                    0

In [631]:
df_test['Name'].value_counts()

Maruti Wagon R LXI CNG               307
Honda City 1.5 S MT                  299
Maruti Swift VDI                     282
Maruti Alto K10 VXI                  260
Hyundai i10 Sportz                   260
                                    ... 
Mahindra Scorpio S10 7 Seater          1
Toyota Etios Liva Diesel               1
Mahindra Bolero SLE BSIII              1
Toyota Etios VXD                       1
Maruti Ertiga VDI Limited Edition      1
Name: Name, Length: 1431, dtype: int64

Null value treatment for all the columns which have null values

In [632]:
df.isnull().sum()

Name                     0
Location                 0
Year                     0
Kilometers_Driven        0
Fuel_Type                0
Owner_Type               0
Mileage                 92
Engine                  37
Power                  797
Seats                   92
Price                    0
Transmission_Manual      0
dtype: int64

In [633]:
np.mean(df['Mileage'])

18.34356167440166

In [634]:
df['Mileage'].fillna(np.mean(df['Mileage']),inplace=True)

In [635]:
np.mean(df['Engine'])

1637.3349953256466

In [636]:
df['Engine'].fillna(np.mean(df['Engine']),inplace=True)

In [637]:
df['Seats'].fillna(df.groupby(['Name'])['Seats'].mean(),inplace=True)

In [638]:
df['Power'].fillna(df.groupby(['Name'])['Power'].mean(),inplace=True)

In [639]:
round(df['Seats'].mean())

5

In [640]:
df['Seats'].fillna(round(df['Seats'].mean()),inplace=True)

In [641]:
(df['Power'].mean())

117.01789729530847

In [642]:
df['Power'].fillna(df['Power'].mean(),inplace=True)

In [643]:
df.isnull().sum()

Name                   0
Location               0
Year                   0
Kilometers_Driven      0
Fuel_Type              0
Owner_Type             0
Mileage                0
Engine                 0
Power                  0
Seats                  0
Price                  0
Transmission_Manual    0
dtype: int64

In [644]:
df_test.isnull().sum()

Name                      0
Location                  0
Year                      0
Kilometers_Driven         0
Fuel_Type                 0
Owner_Type                0
Mileage                 249
Engine                   69
Power                  1836
Seats                   246
Transmission_Manual       0
dtype: int64

In [645]:
np.mean(df_test['Mileage'])

18.272973340222563

In [646]:
df_test['Mileage'].fillna(np.mean(df_test['Mileage']),inplace=True)

In [647]:
np.mean(df_test['Engine'])

1643.5527264404645

In [648]:
df_test['Engine'].fillna(np.mean(df_test['Engine']),inplace=True)

In [649]:
round(df_test['Seats'].mean())

5

In [650]:
df_test['Seats'].fillna(round(df_test['Seats'].mean()),inplace=True)

In [651]:
(df_test['Power'].mean())

116.60467797646835

In [652]:
df_test['Power'].fillna(df_test['Power'].mean(),inplace=True)

In [653]:
df_test.isnull().sum()

Name                   0
Location               0
Year                   0
Kilometers_Driven      0
Fuel_Type              0
Owner_Type             0
Mileage                0
Engine                 0
Power                  0
Seats                  0
Transmission_Manual    0
dtype: int64

In [654]:
df_test.dtypes

Name                    object
Location               float64
Year                     int64
Kilometers_Driven        int64
Fuel_Type                int64
Owner_Type               int64
Mileage                float64
Engine                 float64
Power                  float64
Seats                  float64
Transmission_Manual      uint8
dtype: object

In [655]:
from sklearn.model_selection import train_test_split

In [656]:
y = df['Price']
x = df.drop(columns=['Price'])
train_x,test_x,train_y,test_y = train_test_split(x,y,train_size=0.8,random_state=2)
print(train_x.shape,train_y.shape,test_x.shape,test_y.shape)

(18000, 11) (18000,) (4500, 11) (4500,)


In [657]:
#OLS Model

In [658]:
import statsmodels.api as sm

In [659]:
train_x_c = sm.add_constant(train_x)
test_x_c = sm.add_constant(test_x)
base_model = sm.OLS(train_y,train_x_c).fit()
base_model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  Price   R-squared:                       0.230
Model:                            OLS   Adj. R-squared:                  0.229
Method:                 Least Squares   F-statistic:                     487.2
Date:                Fri, 14 Jul 2023   Prob (F-statistic):               0.00
Time:                        13:35:15   Log-Likelihood:                -58932.
No. Observations:               18000   AIC:                         1.179e+05
Df Residuals:                   17988   BIC:                         1.180e+05
Df Model:                          11                                         
Covariance Type:            nonrobust                                         
=======================================================================================
                          coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------------
const                -158.4513     37.312     -4.247      0.000    -231.587     -85.316
Name                 5708.0329    184.280     30.975      0.000    5346.826    6069.239
Location               -6.6331      0.866     -7.658      0.000      -8.331      -4.935
Year                    0.0799      0.019      4.309      0.000       0.044       0.116
Kilometers_Driven   -1.247e-05    2.3e-06     -5.429      0.000    -1.7e-05   -7.97e-06
Fuel_Type              -1.2332      0.054    -22.845      0.000      -1.339      -1.127
Owner_Type             -0.6892      0.116     -5.940      0.000      -0.917      -0.462
Mileage                -0.0362      0.010     -3.799      0.000      -0.055      -0.018
Engine                  0.0009   7.53e-05     12.169      0.000       0.001       0.001
Power                   0.0055      0.001      7.210      0.000       0.004       0.007
Seats                   1.0415      0.059     17.611      0.000       0.926       1.157
Transmission_Manual    -3.3057      0.127    -25.999      0.000      -3.555      -3.057
==============================================================================
Omnibus:                    10870.491   Durbin-Watson:                   2.023
Prob(Omnibus):                  0.000   Jarque-Bera (JB):           132789.614
Skew:                           2.727   Prob(JB):                         0.00
Kurtosis:                      15.137   Cond. No.                     2.08e+08
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 2.08e+08. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [660]:
from sklearn.metrics import r2_score , mean_squared_error
from sklearn.linear_model import LinearRegression

In [661]:
train_x,test_x,train_y,test_y = train_test_split(x,y,train_size=0.8,random_state=2)

lr = LinearRegression()
model = lr.fit(train_x,train_y)

pred_train = model.predict(train_x)
pred_test= model.predict(test_x)

print("RMSE Train :", np.sqrt(mean_squared_error(train_y,
                                                 pred_train)))

print("RMSE Test  :", np.sqrt(mean_squared_error(test_y,
                                                 pred_test)))

print("R Square Train :", (r2_score(train_y,
                                                 pred_train)))

print("R Square Test  :", (r2_score(test_y,
                                                 pred_test)))

RMSE Train : 6.392063087480955
RMSE Test  : 6.56405596127134
R Square Train : 0.2295291367425002
R Square Test  : 0.20967175849525854


In [662]:
df['Price']

0        25.53
1         4.58
2         2.37
3         1.20
4        33.04
         ...  
22495     2.23
22496     2.36
22497     2.58
22498    16.46
22499     4.92
Name: Price, Length: 22500, dtype: float64

frequency_ = df_test['Name'].value_counts(normalize=True)
df_test['Name'] = df_test['Name'].map(frequency_)

In [672]:
test_x_c= sm.add_constant(df_test)

In [664]:
df_test.dtypes

Name                    object
Location               float64
Year                     int64
Kilometers_Driven        int64
Fuel_Type                int64
Owner_Type               int64
Mileage                float64
Engine                 float64
Power                  float64
Seats                  float64
Transmission_Manual      uint8
dtype: object

In [665]:
df.head(2)

Name  Location  Year  Kilometers_Driven  Fuel_Type  Owner_Type  \
0  0.000729  0.057867  2012              33715          1           1   
1  0.000850  0.065556  2013              73977          1           1   

   Mileage  Engine  Power  Seats  Price  Transmission_Manual  
0    16.46  1248.0  100.0    5.0  25.53                    1  
1    18.60  2993.0   61.0    4.0   4.58                    1

In [666]:
df_test.head()

Name  Location  Year  Kilometers_Driven  Fuel_Type  \
0     Ford Figo Diesel Titanium  0.065886  2010              48928          1   
1                Maruti SX4 VDI  0.080362  2011              47428          1   
2   Hyundai Creta 1.6 SX Option  0.058514  2014              28984          3   
3  Jeep Compass 2.0 Limited 4X4  0.049333  2017              46735          3   
4        Maruti Swift Dzire VDI  0.049333  2015              19975          3   

   Owner_Type  Mileage  Engine  Power  Seats  Transmission_Manual  
0           1    17.80  1248.0  258.0    7.0                    0  
1           1    23.90  2498.0  183.0    5.0                    0  
2           1    19.59  2494.0  203.0    5.0                    0  
3           1    23.90  1498.0  215.0    5.0                    1  
4           1    26.60  2993.0  147.8    7.0                    0

In [670]:
df.drop('Name',axis=1,inplace=True)
df_test.drop('Name',axis=1,inplace=True)

In [673]:
test_x_c.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 52500 entries, 0 to 52499
Data columns (total 11 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   const                52500 non-null  float64
 1   Location             52500 non-null  float64
 2   Year                 52500 non-null  int64  
 3   Kilometers_Driven    52500 non-null  int64  
 4   Fuel_Type            52500 non-null  int64  
 5   Owner_Type           52500 non-null  int64  
 6   Mileage              52500 non-null  float64
 7   Engine               52500 non-null  float64
 8   Power                52500 non-null  float64
 9   Seats                52500 non-null  float64
 10  Transmission_Manual  52500 non-null  uint8  
dtypes: float64(6), int64(4), uint8(1)
memory usage: 4.1 MB


In [674]:
y=model.predict(test_x_c)

C:\Users\admin\anaconda3\lib\site-packages\sklearn\base.py:493: FutureWarning: The feature names should match those that were passed during fit. Starting version 1.2, an error will be raised.
Feature names unseen at fit time:
- const
Feature names seen at fit time, yet now missing:
- Name

  warnings.warn(message, FutureWarning)


In [675]:
y

array([5716.37710838, 5714.80801369, 5713.21876518, ..., 5711.32351386,
       5710.76434464, 5710.48542626])

In [677]:
df_sample_solution = pd.read_csv('Sample_Submission.csv')
df_sample_solution['Price']=y

In [679]:
df_sample_solution.to_csv('Solution_File.csv')

In [680]:
from sklearn.metrics import r2_score,mean_squared_error,mean_absolute_percentage_error

In [681]:
def validation(test,pred):
    r2 = r2_score(test,pred)
    print('R-Square: ',r2)
    n = test_x.shape[0]
    k = test_x.shape[1]
    adj_r2= 1-((1-r2)*(n-1)/(n-k-1))
    print('Adjusted R2: ',adj_r2)
    print('MSE: ',mean_squared_error(test,pred))
    print('RMSE: ',mean_squared_error(test,pred)**0.5)
    print('MAPE: ',mean_absolute_percentage_error(test,pred))

In [682]:
def model(model,trainx,trainy,testx,testy):
    print('%s'%model)
    model.fit(trainx,trainy)
    ypred = model.predict(testx)
    validation(testy,ypred)
    print()

In [683]:
from sklearn.linear_model import LinearRegression,Ridge,Lasso,ElasticNet,SGDRegressor

In [684]:
model(LinearRegression(),train_x,train_y,test_x,test_y)

LinearRegression()
R-Square:  0.20967175849525854
Adjusted R2:  0.20773467947196256
MSE:  43.08683066270181
RMSE:  6.56405596127134
MAPE:  1.0096573934242326



In [685]:
model(Ridge(alpha=1),train_x,train_y,test_x,test_y)

Ridge(alpha=1)
R-Square:  0.1746610255691673
Adjusted R2:  0.17263813592595445
MSE:  44.995533201398224
RMSE:  6.707870988726469
MAPE:  1.0176420377119293



In [ ]:
y=model.pr

In [686]:
ridge = Ridge(alpha=0.01,max_iter=500)

ridge.fit(train_x,train_y)

ytrain_pred = ridge.predict(train_x)
rmse_train = mean_squared_error(train_y,ytrain_pred)**0.5

ytest_pred = ridge.predict(test_x)
rmse_test = mean_squared_error(test_y,ytest_pred)**0.5

rmse_test , rmse_train

(6.677127217910843, 6.526450627417337)

In [687]:
test_x_c= sm.add_constant(df_test)

In [688]:
test_x_c.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 52500 entries, 0 to 52499
Data columns (total 11 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   const                52500 non-null  float64
 1   Location             52500 non-null  float64
 2   Year                 52500 non-null  int64  
 3   Kilometers_Driven    52500 non-null  int64  
 4   Fuel_Type            52500 non-null  int64  
 5   Owner_Type           52500 non-null  int64  
 6   Mileage              52500 non-null  float64
 7   Engine               52500 non-null  float64
 8   Power                52500 non-null  float64
 9   Seats                52500 non-null  float64
 10  Transmission_Manual  52500 non-null  uint8  
dtypes: float64(6), int64(4), uint8(1)
memory usage: 4.1 MB


In [690]:
y=ridge.predict(test_x_c)

C:\Users\admin\anaconda3\lib\site-packages\sklearn\base.py:493: FutureWarning: The feature names should match those that were passed during fit. Starting version 1.2, an error will be raised.
Feature names unseen at fit time:
- const
Feature names seen at fit time, yet now missing:
- Name

  warnings.warn(message, FutureWarning)


In [691]:
df_sample_solution = pd.read_csv('Sample_Submission.csv')
df_sample_solution['Price']=y

In [694]:
df_sample_solution.to_csv('Solution_File_1.csv')

In [695]:
model(Lasso(alpha=0.1),train_x,train_y,test_x,test_y)

Lasso(alpha=0.1)
R-Square:  0.1725904523017765
Adjusted R2:  0.17056248772408478
MSE:  45.10841596967299
RMSE:  6.716279920437577
MAPE:  1.025728561156237



In [696]:

enet =ElasticNet(alpha=0.1 , l1_ratio=0.001,max_iter=500)

enet.fit(train_x , train_y)


ytrain_pred = enet.predict(train_x)
rmse_train = mean_squared_error(train_y,ytrain_pred)**0.5

ytest_pred = enet.predict(test_x)
rmse_test = mean_squared_error(test_y,ytest_pred)**0.5

rmse_test , rmse_train

(6.73328593750956, 6.602446482779143)

In [699]:
test_x_c= sm.add_constant(df_test)
y=enet.predict(test_x_c)

C:\Users\admin\anaconda3\lib\site-packages\sklearn\base.py:493: FutureWarning: The feature names should match those that were passed during fit. Starting version 1.2, an error will be raised.
Feature names unseen at fit time:
- const
Feature names seen at fit time, yet now missing:
- Name

  warnings.warn(message, FutureWarning)


In [700]:
df_sample_solution = pd.read_csv('Sample_Submission.csv')
df_sample_solution['Price']=y

In [703]:
df_sample_solution.to_csv('Solution_File_2.csv')

In [704]:
from sklearn.ensemble import RandomForestRegressor 

In [705]:
rf=RandomForestRegressor()    #regressor because target variable is float

# build the model and find the r2score with RMSE

rf = rf.fit(train_x,train_y)

pred_train_rf = rf.predict(train_x)

pred_test_rf=rf.predict(test_x)

print("RMSE Train :", np.sqrt(mean_squared_error(train_y,
                                                 pred_train_rf)))

print("RMSE Test  :", np.sqrt(mean_squared_error(test_y,
                                                 pred_test_rf)))

RMSE Train : 2.431195613820492
RMSE Test  : 6.580037455017004


In [706]:
test_x_c= sm.add_constant(df_test)
y=rf.predict(test_x_c)

C:\Users\admin\anaconda3\lib\site-packages\sklearn\base.py:493: FutureWarning: The feature names should match those that were passed during fit. Starting version 1.2, an error will be raised.
Feature names unseen at fit time:
- const
Feature names seen at fit time, yet now missing:
- Name

  warnings.warn(message, FutureWarning)


In [1]:
df_sample_solution = pd.read_csv('Sample_Submission.csv')
df_sample_solution['Price']=y

NameError: name 'pd' is not defined

In [708]:
df_sample_solution.to_csv('Solution_File_3.csv')